File: 总题库/23.MergeKSortedLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
// 分治
class Solution {
    ListNode* mergeTwoLists(ListNode *a, ListNode *b) {
        if ((!a) || (!b)) return a ? a : b;
        ListNode head, *tail = &head;
        while (a && b) {
            if (a->val < b->val) {
                tail->next = a;
                a = a->next;
            }
            else {
                tail->next = b;
                b = b->next;
            }
            tail = tail->next;
        }
        tail->next = (a ? a : b);
        return head.next;
    }

    ListNode* merge(vector<ListNode*> &lists, int left, int right) {
        if (left == right) return lists[left];
        if (left > right) return NULL;
        int mid = (left + right) >> 1;
        return mergeTwoLists(merge(lists, left, mid), merge(lists, mid + 1, right));
    }
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.size() == 0) {
            return NULL;
        }
        int len = lists.size();
        for (int interval = 1; interval < len; interval *= 2) {
            for (int i = interval; i < len; i += 2 * interval) {
                lists[i - interval] = mergeTwoLists(lists[i - interval], lists[i]);
                lists[i] = NULL;
            }
        }
        return lists[0];
    }
};

File: 总题库/23.MergeKSortedLists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeTwoLists(self, l1: List[ListNode], l2: List[ListNode]) -> ListNode:
        if l1 is None:
            return l2
        if l2 is None:
            return l1
        
        head = ListNode
        tail = head
        while l1 and l2:
            if l1.val < l2.val:
                tail.next = l1
                l1 = l1.next
            else:
                tail.next = l2
                l2 = l2.next
            tail = tail.next
        
        tail.next = l1 if l1 else l2

        return head.next

    def merge(self, lists: List[ListNode], left: int, right: int) -> ListNode:
        if left == right: return lists[left]
        if left > right: return None
        mid = (left + right) >> 1
        return self.mergeTwoLists(self.merge(lists, left, mid), self.merge(lists, mid + 1, right))

    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        return self.merge(lists, 0, len(lists) - 1)